### PR TITLE
Implemented deleting files when updating

### PR DESCRIPTION
**Pull request by NicknineTheEagle**
Fixes https://github.com/SteamRE/DepotDownloader/issues/112.
Implementation is rather basic but I'm open to suggestions.
Also fixed loading old manifest which I broke in the last PR.

### CODE REVIEW

**Comment by NicknineTheEagle** [comment]
@azuisleet It's been 3 months, are there any criticisms worth addressing? Do note that this PR also fixes a regression from https://github.com/SteamRE/DepotDownloader/pull/94.

**Comment by azuisleet** [comment]
I do intend to merge this, but there's another PR outstanding that I will need to merge this with https://github.com/SteamRE/DepotDownloader/pull/132

File: DepotDownloader/ContentDownloader.cs [hunk 1]
```diff
@@ -1004,6 +1005,34 @@ private static async Task DownloadSteam3Async( uint appId, List<DepotDownloadInf
...
+                    var oldfilesAfterExclusions = oldProtoManifest.Files.AsParallel().Where( f => TestIsFileIncluded( f.FileName ) ).ToList();
+
+                    foreach ( var file in oldfilesAfterExclusions )
+                    {
+                        // Delete it if it's in the old manifest AND not in the new manifest AND not in any of the previous depots.
+                        var newManifestFile = filesAfterExclusions.SingleOrDefault( f => f.FileName == file.FileName );
+                        if ( newManifestFile == null )
+                            continue;
+
+                        var previousFile = previousFiles.SingleOrDefault( f => f.FileName == file.FileName );
```
Comment (xPaw) [inline DepotDownloader/ContentDownloader.cs:1020]: This is incorrect to check here, as `previousFiles` is being filled within the same loop.

File: DepotDownloader/ContentDownloader.cs [hunk 2]
```diff
@@ -1004,6 +1005,34 @@ private static async Task DownloadSteam3Async( uint appId, List<DepotDownloadInf
...
+                        if ( previousFiles == null )
```
Comment (xPaw) [inline DepotDownloader/ContentDownloader.cs:1021]:
```suggestion
                        if ( previousFile == null )

```

File: DepotDownloader/ContentDownloader.cs (hunk 1 above)
Comment (NicknineTheEagle) [reply DepotDownloader/ContentDownloader.cs:1020]: It's being filled with external loop after the current depot is handled so this snippet checks files from previous depots. I'm pretty sure this is correct.

**Comment by xPaw** [comment]
So this doesn't handle moving files between depots, right? Dota updates for example sometimes may move vpks between depots.

There's also this game where they put entire game versions into new depots (which is incorrect but still, an example): https://steamdb.info/app/1404090/depots/

**Comment by NicknineTheEagle** [comment]
> 
> 
> So this doesn't handle moving files between depots, right? Dota updates for example sometimes may move vpks between depots.
> 
> There's also this game where they put entire game versions into new depots (which is incorrect but still, an example): https://steamdb.info/app/1404090/depots/

It doesn't although there is a check to make sure it doesn't delete a file that is any of the previous depot.

**Comment by muk-as** [comment]
> > So this doesn't handle moving files between depots, right? Dota updates for example sometimes may move vpks between depots.
> > There's also this game where they put entire game versions into new depots (which is incorrect but still, an example): https://steamdb.info/app/1404090/depots/
> 
> It doesn't although there is a check to make sure it doesn't delete a file that is also in any of the previous depots.

its ok if u download with option -dir, coz all DEPOTs files will downloaded into 1 folder (no duplicate - all ok)
but not ok if u dosnt use -dir, coz each DEPOT will downloaded to own folder (if file from 1 DEPOT will move to another DEPOT then will duplicates)

**Comment by azuisleet** [comment]
That makes sense. I've addressed this in 30312af050cbdb432818628adff7fbd65ccd3cfa.

**Pull request merged**